### PR TITLE
Test with Java 21 and modernization

### PR DESCRIPTION
**Pull request by gounthar**
Test with Java 21
==

Java 21 was released Sep 19, 2023. We want to announce full support for Java 21 in early October and would like the most used plugins to be compiled and tested with Java 21.

The acceptance test harness and plugin bill of materials tests are already passing with Java 21. This is a further step to improve plugin readiness for use with Java 21 and for development with Java 21.

The change intentionally tests only two Java configurations, Java 17 and Java 21 because we believe that the risk of a regression that only affects Java 11 is shallow. We generate Java 11 byte code with the Java 17 and the Java 21 builds, so we're already testing Java 11 byte code.

Testing done
===

Confirmed tests pass with Java 21 after upgrading to a recent parent POM and to Jenkins version 2.361.4 as the minimum version.

### CODE REVIEW

**Comment by gounthar** [comment]
This change is not compatible with JDK8, hence the fail.
We could replay this build while changing the content of the master Jenkinsfile with this one; that should go fine.

**Comment by gounthar** [comment]
We're still using JDK8, and that's why the build is failing.
Could someone please re-run it with our proposed Jenkinsfile?
Thanks.

**Comment by basil** [comment]
https://ci.jenkins.io/job/Plugins/job/pipeline-graph-analysis-plugin/job/PR-121/2/

**Comment by gounthar** [comment]
Thank you, Basil. 
Unfortunately, JDK17 failed. I'll test it on my machine.

**Comment by gounthar** [comment]
JDK17 works on my machine. 🤔

File: pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>4.40</version>
+        <version>4.73</version>
```
Comment (MarkEWaite) [inline pom.xml:8]: Optional, but nice to use most recent parent pom release.
```suggestion
        <version>4.74</version>
```

File: pom.xml [hunk 2]
```diff
@@ -54,8 +54,8 @@
         <dependencies>
             <dependency>
                 <groupId>io.jenkins.tools.bom</groupId>
-                <artifactId>bom-2.303.x</artifactId>
-                <version>1289.v5c4b_1c43511b_</version>
+                <artifactId>bom-2.361.x</artifactId>
+                <version>2120.v89a_c54a_1e4f9</version>
```
Comment (MarkEWaite) [inline pom.xml:58]: Needs to use final release of bom-2.361.x, not an incremental
```suggestion
                <version>2102.v854b_fec19c92</version>
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Needs fix for plugin bom version

**Comment by gounthar** [comment]
Thank you, Mark.

**Review by jglick — APPROVED** [review]

**Pull request merged**